### PR TITLE
Deployment fixes

### PR DESCRIPTION
**Pull request by Uberi**
* Allow management of the test DB via pgAdmin.
* Fix Vagrant config.
* Update Ansible setup to make development environment setup work again.
* Fix `run-tests.sh`.
* Add TravisCI config and make it pass.
* Use TCP/IP connections instead of socket files, which didn't work reliably with Travis..

Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=1249653

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,11 +20,16 @@ To deploy a new version of the aggregation service to the cloud:
 ansible-playbook ansible/deploy.yml -e '@ansible/envs/dev.yml' -i ansible/inventory
 ```
 
+To connect to the database on the host, when running it inside the Vagrant VM, using something like [pgAdmin](http://pgadmin.org/), use `localhost` as the host, `5432` as the port, `vagrant` as the username, and a blank password.
+
 ## API
-Aggregates are made available through a HTTP API. There are two kinds of aggregates: per submission date and per build-id. 
+Aggregates are made available through a HTTP API. There are two kinds of aggregates: per submission date (date a ping is submitted) and per build-id (date the submitting product was built).
```
Comment (vitillo) [inline README.md]: `submission date` is actually the date the ping was received by our servers.

File: ansible/dev.yml [hunk 2]
```diff
@@ -5,18 +5,99 @@
...
+        - wget
+        - ca-certificates
+
+    - name: get the current release codename for adding apt repositories
+      shell: "lsb_release -cs"
+      register: codename
+
+    - name: add apt repository for postgresql
+      sudo: yes
+      apt_repository: repo="deb http://apt.postgresql.org/pub/repos/apt/ {{ codename.stdout }}-pgdg main" state=present
```
Comment (vitillo) [inline ansible/dev.yml]: Why can't we use the packages in the default repositories?

File: ansible/dev.yml [hunk 3]
```diff
@@ -5,18 +5,99 @@
...
+    - name: source anaconda
+      lineinfile: dest={{ansible_env.HOME}}/.bashrc regexp="^export.*miniconda2/bin" line="export PATH={{ansible_env.HOME}}/miniconda2/bin:$PATH"
+
+    - name: install common python packages
+      command: "{{item}} chdir={{ansible_env.HOME}}"
+      with_items:
+        - miniconda2/bin/conda update --yes conda
+        - miniconda2/bin/conda install --yes atlas numpy scipy matplotlib nose dateutil pandas flask supervisor tornado
+
+    - name: setup Python dependencies
```
Comment (vitillo) [inline ansible/dev.yml]: The python dependencies should be installed from `https://github.com/mozilla/python_mozaggregator/blob/master/setup.py#L18` so that we have an unique definition.

File: ansible/dev.yml [hunk 4]
```diff
@@ -5,18 +5,99 @@
...
+        - ./miniconda.sh -b -f
+
+    - name: source anaconda
+      lineinfile: dest={{ansible_env.HOME}}/.bashrc regexp="^export.*miniconda2/bin" line="export PATH={{ansible_env.HOME}}/miniconda2/bin:$PATH"
+
+    - name: install common python packages
+      command: "{{item}} chdir={{ansible_env.HOME}}"
+      with_items:
+        - miniconda2/bin/conda update --yes conda
+        - miniconda2/bin/conda install --yes atlas numpy scipy matplotlib nose dateutil pandas flask supervisor tornado
```
Comment (vitillo) [inline ansible/dev.yml]: See comment below about python dependencies.

File: ansible/dev.yml [hunk 5]
```diff
@@ -5,18 +5,99 @@
...
+        - flask-sslify
+        - gevent
+        - psycogreen
+        - ujson
+        - nose
+
+    # we have to use wget here instead of the built in Ansible get_url module
+    # because Python versions before 2.7.9 don't support SNI, which is needed
+    # to verify the download via HTTPS, while wget does
+    - name: get spark
```
Comment (vitillo) [inline ansible/dev.yml]: Spark stuff should be defined in its own ansible role.

File: ansible/dev.yml [hunk 6]
```diff
@@ -5,18 +5,99 @@
...
+      with_items:
+        - libpq-dev
+        - postgresql-{{postgresql_version}}
+        - postgresql-contrib-{{postgresql_version}}
+        - postgresql-server-dev-{{postgresql_version}}
+
+    # we have to use wget here instead of the built in Ansible get_url module
+    # because Python versions before 2.7.9 don't support SNI, which is needed
+    # to verify the download via HTTPS, while wget does
+    - name: install anaconda
```
Comment (vitillo) [inline ansible/dev.yml]: anaconda stuff should be defined in its own ansible role.

File: run-tests.sh [hunk 7]
```diff
@@ -2,60 +2,64 @@
...
 
 if [ -z "$SPARK_HOME" ]; then
-    echo 'You need to set $SPARK_HOME to run these tests.' >&2
-    exit 1
+	echo 'You need to set $SPARK_HOME to run these tests.' >&2
+	exit 1
 fi
 
-PYFORJ=`ls -1 $SPARK_HOME/python/lib/py4j-*-src.zip | head -1`
+# make sure pyspark, py4j, and mozaggregator are importable
```
Comment (vitillo) [inline run-tests.sh]: How does this comment apply to mozaggregator as well?

File: run-tests.sh [hunk 8]
```diff
@@ -2,60 +2,64 @@
...
 
 clean_exit() {
-    local error_code="$?"
-    for job in $(jobs -p); do
-	pkill -9 -P $job >/dev/null 2>&1 || true
-	kill -9 $job >/dev/null 2>&1 || true
-    done
-    rm -rf "PGSQL_DATA"
-    return $error_code
+	local error_code="$?" # save the original error code
```
Comment (vitillo) [inline run-tests.sh]: Why was the indentation length changed?

File: run-tests.sh [hunk 9]
```diff
@@ -2,60 +2,64 @@
...
-PGSQL_PATH=`pg_config --bindir`
-${PGSQL_PATH}/initdb ${PGSQL_DATA}
-echo "host all all 0.0.0.0/0 trust" >> $PGSQL_DATA/pg_hba.conf
-mkfifo ${PGSQL_DATA}/out
-${PGSQL_PATH}/postgres -h '*' -F -k ${PGSQL_DATA} -D ${PGSQL_DATA} &> ${PGSQL_DATA}/out &
-# Wait for PostgreSQL to start listening to connection
-wait_for_line "database system is ready to accept connections" ${PGSQL_DATA}/out
-export DB_TEST_URL="postgresql:///?host=${PGSQL_DATA}&dbname=template1"
+# make sure we have everything we need to run
+check_for_cmd ${PGSQL_PATH}/initdb
```
Comment (vitillo) [inline run-tests.sh]: Do we really need all these checks?

File: tests/test_db.py [hunk 10]
```diff
@@ -13,16 +13,18 @@
...
     global aggregates
     global sc
 
-    logger = logging.getLogger("py4j")
-    logger.setLevel(logging.ERROR)
-
     sc = pyspark.SparkContext(master="local[*]")
     raw_pings = list(generate_pings())
     aggregates = _aggregate_metrics(sc.parallelize(raw_pings))
+    submit_aggregates(aggregates)
```
Comment (vitillo) [inline tests/test_db.py]: Why was this line moved to `setup_module`?

File: ansible/dev.yml (hunk 2 above)
Comment (Uberi) [inline ansible/dev.yml]: The postgres version available in Ubuntu Trusty (which is the latest version supported by Travis), is only 9.1, while we're using 9.4 features.

(That's why the Vagrant image was switched to Trusty, and also the fact that it's LTS.)

File: run-tests.sh (hunk 7 above)
Comment (Uberi) [inline run-tests.sh]: `mozaggregator` needs to be imported from the current directory, but if we use relative imports, processes created by Spark will in certain cases fail to pick up the relative path when deserializing. This happens especially often on Travis, for some reason. I'll add a comment.

File: run-tests.sh (hunk 9 above)
Comment (Uberi) [inline run-tests.sh]: Not really, but if things go wrong the error messages tend to be really cryptic, and it's pretty simple to test for. I ran into exactly that issue when the binary directory was set wrong on Travis.

File: tests/test_db.py (hunk 10 above)
Comment (Uberi) [inline tests/test_db.py]: Tests can be run out of order, which makes the other tests, which rely on the aggregates being submitted already, fail if they run before `test_submit` does.

Moving it into the setup routine guarantees that those tests will have aggregates to test against. Since `test_submit` isn't checking the first submission, this doesn't affect the test behaviour.

**Comment by Uberi** [comment]
Changes:

* TravisCI was broken by using distutils easy_install, fixed by using Pip.
* Add some more comments.

File: ansible/dev.yml [hunk 11]
```diff
@@ -25,16 +48,21 @@
...
     - role: spark
       spark_version: "{{spark_version}}"
 
   post_tasks:
-    - name: setup project dependencies on vagrant
-      command: "{{ansible_env.HOME}}/miniconda2/bin/python setup.py develop chdir=/vagrant"
+    # travis_build_dir can be set on the command line with Ansible's `-e` flag
+
+    - name: setup project dependencies on Vagrant
+      pip: name=/vagrant state=present executable={{ansible_env.HOME}}/miniconda2/bin/pip
```
Comment (vitillo) [inline ansible/dev.yml]: The nice thing about using `python setup.py develop` was that one could change the source files and re-run the tests without having to install the package first. Is there a way to do the same with pip?

File: ansible/roles/spark/tasks/main.yml [hunk 12]
```diff
@@ -20,5 +23,4 @@
   lineinfile: >
     dest={{ansible_env.HOME}}/.bashrc
     regexp="^export PYTHONPATH"
-    line="export PYTHONPATH=$SPARK_HOME/python"
-  
\ No newline at end of file
+    line="export PYTHONPATH=$SPARK_HOME/python:$SPARK_HOME/python/lib/py4j-0.8.2.1-src.zip"
```
Comment (vitillo) [inline ansible/roles/spark/tasks/main.yml]: Why is the py4j zip part of the PYTHONPATH? What if the version of py4j changes in a newer Spark release?

File: run-tests.sh [hunk 13]
```diff
@@ -6,56 +6,62 @@ if [ -z "$SPARK_HOME" ]; then
     exit 1
 fi
 
-PYFORJ=`ls -1 $SPARK_HOME/python/lib/py4j-*-src.zip | head -1`
+# make sure pyspark, py4j, and mozaggregator are importable
+# mozaggregator needs to be imported from the current directory
+# this mainly is needed for TravisCI support
+PYFORJ=$(ls -1 $SPARK_HOME/python/lib/py4j-*-src.zip | head -1)
+export PYTHONPATH=$PYTHONPATH:$SPARK_HOME/python:$PYFORJ:.
```
Comment (vitillo) [inline run-tests.sh]: Since we define the `PYTHONPATH` variable in .bashrc, could we just source that file?

File: run-tests.sh [hunk 14]
```diff
@@ -6,56 +6,62 @@ if [ -z "$SPARK_HOME" ]; then
...
+echo "host all all 0.0.0.0/0 trust" >> $PGSQL_DATA/pg_hba.conf # allow anyone to access the database
+mkfifo ${PGSQL_DATA}/out # create output FIFO for the database to let us read the output programmatically
+(${PGSQL_PATH}/postgres -h '*' -F -k ${PGSQL_DATA} -D ${PGSQL_DATA} 2>&1 | tee ${PGSQL_DATA}/out &) # start the database server
+wait_for_line "database system is ready to accept connections" ${PGSQL_DATA}/out # wait for PostgreSQL to start listening for connections
 
-# Launch db service
+# launch the HTTP API service in the background
 mkfifo ${PGSQL_DATA}/out_service
-python ./mozaggregator/service.py -d &> ${PGSQL_DATA}/out_service &
+(~/miniconda2/bin/python ./mozaggregator/service.py -d 2>&1 | tee ${PGSQL_DATA}/out_service &)
```
Comment (vitillo) [inline run-tests.sh]: Could we add ~/miniconda2/bin to `PATH`?

Comment (Uberi) [inline run-tests.sh]: That might be a bit confusing - this way makes it obvious that things like `apt-get install python-pandas` won't install Pandas for the Python we're using.

**Comment by Uberi** [comment]
Changes:

* Source bashrc to avoid duplicating variable definitions.
* Make installation editable in dev environment.
* Use the first available py4j version instead of hardcoding..

File: run-tests.sh (hunk 13 above)
Comment (vitillo) [inline run-tests.sh]: The above comment still applies.

File: run-tests.sh [hunk 15]
```diff
@@ -6,56 +6,62 @@ if [ -z "$SPARK_HOME" ]; then
...
     done
     rm -rf "PGSQL_DATA"
     return $error_code
 }
 
 check_for_cmd () {
-    if ! which "$1" >/dev/null 2>&1
-    then
-        echo "Could not find $1 command" 1>&2
+    which "$1" >/dev/null 2>/dev/null || {
```
Comment (vitillo) [inline run-tests.sh]: Could you redirect the standard error to the standard output?

File: run-tests.sh (hunk 14 above)
Comment (vitillo) [inline run-tests.sh]: OK

**Comment by vitillo** [comment]
I am getting the following error with `vagrant up`:

```
...
TASK [spark : install java] ****************************************************
changed: [dev]

TASK [spark : get spark] *******************************************************
fatal: [dev]: FAILED! => {"failed": true, "msg": "recursive loop detected in template string: {{spark_version}}"}

NO MORE HOSTS LEFT *************************************************************
	to retry, use: --limit @ansible/dev.retry

PLAY RECAP *********************************************************************
dev                        : ok=13   changed=12   unreachable=0    failed=1

Ansible failed to complete successfully. Any error output should be
visible above. Please fix these errors and try again.
```

**Comment by Uberi** [comment]
This doesn't seem to happen on my freshly destroyed/rebuilt VM. Maybe this is caused by an old version Ansible?

**Comment by vitillo** [comment]
I have 2.0.1.0.

**Comment by Uberi** [comment]
Seems like Ansible 2 changed some variable resolution behaviour. This should fix things for both 1 and 2.

**Comment by vitillo** [comment]
This time around the ansible deployment worked but when I tried to run the tests for the second time I got the following error:
```bash
vagrant@vagrant-ubuntu-trusty-64:/vagrant$ ./run-tests.sh
The files belonging to this database system will be owned by user "vagrant".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.UTF-8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /tmp/PGSQL-UH97P ... ok
creating subdirectories ... ok
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting dynamic shared memory implementation ... posix
creating configuration files ... ok
creating template1 database in /tmp/PGSQL-UH97P/base/1 ... ok
initializing pg_authid ... ok
initializing dependencies ... ok
creating system views ... ok
loading system objects' descriptions ... ok
creating collations ... ok
creating conversions ... ok
creating dictionaries ... ok
setting privileges on built-in objects ... ok
creating information schema ... ok
loading PL/pgSQL server-side language ... ok
vacuuming database template1 ... ok
copying template1 to template0 ... ok
copying template1 to postgres ... ok
syncing data to disk ... ok

WARNING: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    /usr/lib/postgresql/9.4/bin/postgres -D /tmp/PGSQL-UH97P
or
    /usr/lib/postgresql/9.4/bin/pg_ctl -D /tmp/PGSQL-UH97P -l logfile start

Waiting for 'database system is ready to accept connections' to appear in file '/tmp/PGSQL-UH97P/out'...
LOG:  could not bind IPv4 socket: Address already in use
HINT:  Is another postmaster already running on port 5432? If not, wait a few seconds and retry.
LOG:  could not bind IPv6 socket: Address already in use
HINT:  Is another postmaster already running on port 5432? If not, wait a few seconds and retry.
WARNING:  could not create listen socket for "*"
FATAL:  could not create any TCP/IP sockets
ERROR: waiting for 'database system is ready to accept connections' to appear in file '/tmp/PGSQL-UH97P/out' timed out
```

**Comment by Uberi** [comment]
Changes:

* Make multiple test runs work properly again - properly kill the service when exiting.

**Comment by vitillo** [comment]
Looks good, thanks. Could you please rebase your changes?

**Comment by Uberi** [comment]
Yep, rebased.
